idea/src/org/jetbrains/kotlin/idea/inspections/MemberVisibilityCanPrivateInspection.kt

```diff
@@ -0,0 +1,89 @@
+import com.intellij.psi.PsiElementVisitor
+import com.intellij.psi.PsiReference
+import com.intellij.psi.search.LocalSearchScope
+import com.intellij.psi.search.searches.ReferencesSearch
+import com.intellij.util.Processor
+import org.jetbrains.kotlin.idea.quickfix.AddModifierFix
+import org.jetbrains.kotlin.idea.refactoring.isConstructorDeclaredProperty
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.containingClass
+import org.jetbrains.kotlin.psi.psiUtil.getParentOfType
+import org.jetbrains.kotlin.psi.psiUtil.isInheritable
+import org.jetbrains.kotlin.psi.psiUtil.isOverridable
+
+class MemberVisibilityCanPrivateInspection : AbstractKotlinInspection() {
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
+        return object : KtVisitorVoid() {
+            override fun visitProperty(property: KtProperty) {
+                super.visitProperty(property)
+                if (canBePrivate(property)) {
+                    registerProblem(holder, property)
+                }
+            }
+
+            override fun visitNamedFunction(function: KtNamedFunction) {
+                super.visitNamedFunction(function)
+                if (canBePrivate(function)) {
+                    registerProblem(holder, function)
+                }
+            }
+
+            override fun visitParameter(parameter: KtParameter) {
+                super.visitParameter(parameter)
+                if (parameter.isConstructorDeclaredProperty() && canBePrivate(parameter)) {
+                    registerProblem(holder, parameter)
+                }
+            }
+        }
+    }
+
+
+    private fun canBePrivate(declaration: KtDeclaration): Boolean {
+        if (declaration.hasModifier(KtTokens.PRIVATE_KEYWORD) || declaration.hasModifier(KtTokens.OVERRIDE_KEYWORD)) return false
+        val klass = declaration.containingClass() ?: return false
+        if (!klass.isInheritable() && declaration.hasModifier(KtTokens.PROTECTED_KEYWORD)) return false //reported by ProtectedInFinalInspection
+        if (declaration.isOverridable()) return false
+        var otherUsageFound = false
+        var inClassUsageFound = false
+        ReferencesSearch.search(declaration, declaration.useScope).forEach(Processor<PsiReference> {
+            val usage = it.element
+            if (klass != usage.getParentOfType<KtClass>(false)) {
+                otherUsageFound = true
+                false
+            } else {
+                inClassUsageFound = true
+                true
+            }
+        })
+        return inClassUsageFound && !otherUsageFound
+    }
+
+    private fun registerProblem(holder: ProblemsHolder, declaration: KtDeclaration) {
+        val modifierListOwner = declaration.getParentOfType<KtModifierListOwner>(false) ?: return
+        holder.registerProblem(declaration,
+                               "Can have 'private' visibility",
```

Comment:
Better to include name of declaration in the message, otherwise it's hard to understand results of a batch inspection.